core/src/sumeragi/fault.rs

```diff
@@ -39,865 +37,1333 @@ impl FaultInjection for NoFault {
+                read_index += 1;
+                write_index += 1;
+                continue;
+            }
+            read_index += 1;
+        }
+        transaction_cache.truncate(write_index);
     }
 }
 
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> Handler<NetworkMessage>
-    for SumeragiWithFault<G, F>
+fn request_view_change<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    state_machine_guard: &mut SumeragiStateMachineData,
+    view_change_proof_chain: &mut Vec<Proof>,
+    current_view_change_index: u64,
+) where
+    F: FaultInjection,
 {
-    type Result = ();
+    let mut suspect_proof = Proof {
+        latest_block_hash: state_machine_guard.latest_block_hash,
+        view_change_index: current_view_change_index,
+        signatures: Vec::new(),
+    };
+    suspect_proof
+        .sign(sumeragi.key_pair.clone())
+        .expect("must be able to perform signing");
+
+    view_change_proof_chain.insert_proof(
+        &state_machine_guard
+            .current_topology
+            .sorted_peers()
+            .iter()
+            .cloned()
+            .collect(),
+        state_machine_guard.current_topology.max_faults(),
+        &state_machine_guard.latest_block_hash,
+        &suspect_proof,
+    );
+
+    sumeragi.broadcast_msg(
+        Message::ViewChangeSuggested(ViewChangeSuggested::new(view_change_proof_chain.clone())),
+        &state_machine_guard.current_topology,
+    );
+}
 
-    async fn handle(&mut self, msg: NetworkMessage) -> Self::Result {
-        use NetworkMessage::*;
+#[instrument(skip(initial_latest_block, initial_block_height))]
+pub fn run_sumeragi_main_loop<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    initial_latest_block: HashOf<VersionedCommittedBlock>,
+    initial_block_height: u64,
+) where
+    F: FaultInjection,
+{
+    use std::sync::mpsc::TryRecvError;
+
+    let mut incoming_message_receiver = sumeragi
+        .incoming_message_receiver
+        .lock()
+        .expect("lock on reciever");
+
+    if initial_block_height != 0 && initial_latest_block != Hash::zeroed().typed() {
+        // Normal startup
+
+        let mut state_machine_guard = sumeragi
+            .sumeragi_state_machine_data
+            .lock()
+            .expect("take lock");
+
+        state_machine_guard.latest_block_hash = initial_latest_block;
+        state_machine_guard.latest_block_height = initial_block_height;
+        state_machine_guard.current_topology = Topology::builder()
+            .at_block(state_machine_guard.latest_block_hash)
+            .with_peers(
+                state_machine_guard
+                    .wsv
+                    .peers()
+                    .iter()
+                    .map(|peer| peer.id().clone())
+                    .collect(),
+            )
+            .build(0)
+            .expect("Should be able to reconstruct topology from wsv");
+    } else {
+        // We need to perform a round of some form
```

Comment:
maybe? 

```suggestion
        // We need to perform genesis round
```

Reply:
I said of some form because the genesis round can have already passed and you just recieve the block.